Documentation/guides/D8andR8.md

```diff
@@ -0,0 +1,270 @@
+   in `$(IntermediateOutputPath)android\bin`. If `multidex` is
+   enabled, a `classes2.dex` (and potentially more) are also generated
+   in this location.
+
+# What will this process look like with D8 / R8?
+
+Two new MSBuild tasks named `R8` and `D8` will be created.
+
+1. The [Javac](https://github.com/xamarin/xamarin-android/blob/221a2190ebb3aaec9ecd9b1cf8f7f6174c43153a/src/Xamarin.Android.Build.Tasks/Tasks/Javac.cs)
+   MSBuild task will remain unchanged.
+2. `D8` will run if `$(AndroidEnableMultiDex)` is `False`,
+   `$(AndroidLinkTool)` is not `r8`, and "desugar" by default.
+3. Otherwise, `R8` will run if `$(AndroidEnableMultiDex)` is `True` or
+   `$(AndroidLinkTool)` is `r8` and will also "desugar" by default.
+
+So in addition to be being faster in general (if Google's claims are
+true), we will be calling a *single* command line tool to produce dex
+files!
+
+# So how do developers use it? What are sensible MSBuild property defaults?
+
+Currently, a `csproj` file might have the following properties:
+```xml
+<Project>
+    <PropertyGroup>
+        <AndroidEnableProguard>True</AndroidEnableProguard>
+        <AndroidEnableMultiDex>True</AndroidEnableMultiDex>
+        <AndroidEnableDesugar>True</AndroidEnableDesugar>
+    </PropertyGroup>
+</Project>
+```
+
+To enable the new behavior, we should introduce two new enum-style
+properties:
+- `$(AndroidDexGenerator)` - supports `dx` or `d8`
+- `$(AndroidLinkTool)` - supports `proguard` or `r8`
+
+But for an existing project, a developer could opt-in to the new
+behavior with two properties:
+```xml
+<Project>
+    <PropertyGroup>
+        <AndroidEnableProguard>True</AndroidEnableProguard>
+        <AndroidEnableMultiDex>True</AndroidEnableMultiDex>
+        <AndroidEnableDesugar>True</AndroidEnableDesugar>
+        <!--New properties-->
+        <AndroidDexGenerator>d8</AndroidDexGenerator>
+        <AndroidLinkTool>r8</AndroidLinkTool>
+    </PropertyGroup>
+</Project>
+```
+
+There should be two new MSBuild properties to configure here, because:
+- You could use `D8` in combination with `proguard`, as `R8` is not
+  "feature complete" in comparison to `proguard`.
+- You may not want to use code shrinking at all, but still use `D8`
+  instead of `dx`.
+- You shouldn't be able to use `dx` in combination with `R8`, it
+  doesn't make sense.
+- Developers should be able to use the existing properties for
+  enabling code shrinking, `multidex`, and `desugar`.
+
+Our reasonable defaults would be:
+- If `AndroidDexGenerator` is omitted, `dx` and `CompileToDalvik`
+  should be used. Until D8/R8 integration is deemed stable and enabled
+  by default.
+- If `AndroidDexGenerator` is `d8` and `AndroidEnableDesugar` is
+  omitted, `AndroidEnableDesugar` should be enabled.
+- If `AndroidLinkTool` is omitted and `AndroidEnableProguard` is
+  `true`, we should default `AndroidLinkTool` to `proguard`.
+
+MSBuild properties default to something like:
```

Comment:
This is public facing documentation.  How specific can we be?  Can we avoid "something like"?